File: java/core/src/core/util/Base64.java
/**
 * Author: Timothy Prepscius
 * License: GPLv3 Affero + keep my name in the code!
 */

package core.util;

import core.callback.Callback;
import core.callback.CallbackDefault;

public class Base64
{
	public static String encode(byte[] bytes)
	{
		return Strings.toString(org.bc.util.encoders.Base64.encode(bytes));
	}
	
	public static byte[] decode(String b64)
	{
		return org.bc.util.encoders.Base64.decode(Strings.toBytes(b64));
	}

	public static byte[] encodeBytes(byte[] bytes)
	{
		return org.bc.util.encoders.Base64.encode(bytes);
	}

	public static byte[] decodeBytes(byte[] b64)
	{
		return org.bc.util.encoders.Base64.decode(b64);
	}

	public static Callback decode_()
	{
		return new CallbackDefault() {
			public void onSuccess(Object... arguments) throws Exception {
				next(Base64.decode((String)arguments[0]));
			}
		};
	}
	
	public static Callback encode_()
	{
		return new CallbackDefault() {
			public void onSuccess(Object... arguments) throws Exception {
				next(Base64.encode((byte[])arguments[0]));
			}
		};
	}
	
	public static Callback decodeBytes_()
	{
		return new CallbackDefault() {
			public void onSuccess(Object... arguments) throws Exception {
				next(Base64.decodeBytes((byte[])arguments[0]));
			}
		};
	}
	
	public static Callback encodeBytes_()
	{
		return new CallbackDefault() {
			public void onSuccess(Object... arguments) throws Exception {
				next(Base64.encodeBytes((byte[])arguments[0]));
			}
		};
	}
}

File: cpp/core/mailiverse/core/crypt/Base64.cpp
/**
 * Author: Timothy Prepscius
 * License: BSD + keep my name in the code!
 */

#include "Base64.h"
#include <botan/botan.h>
#include "CryptorUtil.h"

namespace mailiverse {
namespace core {
namespace crypt {

std::string Base64::encode (const Block &block)
{
	Botan::Pipe pipe(new Botan::Base64_Encoder);
	pipe.process_msg(block.data(), block.size());
	Botan::SecureVector<Botan::byte> sv;

	return pipe.read_all_as_string();
}

Block Base64::decode (const std::string &block)
{
	Botan::Pipe pipe(new Botan::Base64_Decoder);
	pipe.process_msg((Botan::byte *)block.data(), block.size());

	return read_all_as_block(&pipe);
}

Block Base64::encodeBytes (const Block &block)
{
	Botan::Pipe pipe(new Botan::Base64_Encoder);
	pipe.process_msg(block.data(), block.size());
	Botan::SecureVector<Botan::byte> sv;

	return read_all_as_block(&pipe);
}

Block Base64::decodeBytes (const Block &block)
{
	Botan::Pipe pipe(new Botan::Base64_Decoder);
	pipe.process_msg((Botan::byte *)block.data(), block.size());

	return read_all_as_block(&pipe);
}

} /* namespace crypt */
} /* namespace core */
} /* namespace mailiverse */
